lib/hamster/vector.rb

```diff
@@ -526,21 +526,9 @@ def shuffle
     # @return [Vector]
     def uniq(&block)
       array = self.to_a
-      if block_given?
-        if array.frozen?
-          self.class.new(array.uniq(&block).freeze)
-        elsif array.uniq!(&block) # returns nil if no changes were made
-          self.class.new(array.freeze)
-        else
-          self
-        end
-      elsif array.frozen?
-        self.class.new(array.uniq.freeze)
-      elsif array.uniq! # returns nil if no changes were made
-        self.class.new(array.freeze)
-      else
-        self
-      end
+      return self.class.new(array.uniq(&block).freeze) if array.frozen?
+      return self.class.new(array.freeze) if array.uniq!(&block)
```

Comment:
Another thing to note is that `Vector#to_a` sometimes returns a frozen array, sometimes not. If it's not frozen, `uniq!` can be used on it to avoid an extra array copy. `self.class.new(to_a.uniq(&block))` is indeed simple, but it always does an extra copy, even when that's not necessary.